File: Tests/Source/LiftTester.cpp
#include <catch2/catch.hpp>
#include "Aspen/Constant.hpp"
#include "Aspen/Lift.hpp"
#include "Aspen/Queue.hpp"

using namespace Aspen;

namespace {
  int no_parameter_function() {
    return 512;
  }

  int no_parameter_function_throw() {
    throw std::runtime_error("");
  }

  int square(int x) {
    return x * x;
  }
}

TEST_CASE("test_lift_no_parameters", "[Lift]") {
  auto reactor = Lift(no_parameter_function);
  REQUIRE(reactor.commit(0) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 512);
}

TEST_CASE("test_lift_no_parameters_throw", "[Lift]") {
  auto reactor = Lift(no_parameter_function_throw);
  REQUIRE(reactor.commit(0) == State::COMPLETE_EVALUATED);
  REQUIRE_THROWS_AS(reactor.eval(), std::runtime_error);
}

TEST_CASE("test_lift_constant_argument", "[Lift]") {
  auto reactor = Lift(square, Constant(5));
  REQUIRE(reactor.commit(0) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 25);
}

TEST_CASE("test_lift_one_argument_updates", "[Lift]") {
  auto queue = Queue<int>();
  auto reactor = Lift(square, &queue);
  REQUIRE(reactor.commit(0) == State::EMPTY);
  queue.push(10);
  REQUIRE(reactor.commit(1) == State::EVALUATED);
  REQUIRE(reactor.eval() == 100);
  REQUIRE(reactor.commit(2) == State::NONE);
  REQUIRE(reactor.eval() == 100);
  queue.push(5);
  REQUIRE(reactor.commit(3) == State::EVALUATED);
  REQUIRE(reactor.eval() == 25);
  REQUIRE(reactor.commit(4) == State::NONE);
  REQUIRE(reactor.eval() == 25);
  queue.set_complete(4);
  REQUIRE(reactor.commit(5) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 16);
}

TEST_CASE("test_complete_arguments", "[Lift]") {
  auto queue = Queue<int>();
  queue.push(10);
  queue.commit(0);
  queue.set_complete();
  auto reactor = Lift(square, &queue);
  REQUIRE(reactor.commit(1) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 100);
}

File: Source/PythonTests/lift_tester.py
import unittest

from fixtures import aspen


def failing(value):
  raise ValueError('evaluation failed')


class TestLift(unittest.TestCase):
  def test_no_arguments(self):
    reactor = aspen.lift(lambda: 512)
    self.assertEqual(reactor.commit(0), aspen.State.COMPLETE_EVALUATED)
    self.assertEqual(reactor.eval(), 512)

  def test_argument(self):
    reactor = aspen.lift(lambda value: value * 2, aspen.constant(21))
    self.assertEqual(reactor.commit(0), aspen.State.COMPLETE_EVALUATED)
    self.assertEqual(reactor.eval(), 42)

  def test_state_value(self):
    reactor = aspen.lift(lambda state: state,
      aspen.StateReactor(aspen.constant(1)))
    self.assertTrue(aspen.has_evaluation(reactor.commit(0)))
    self.assertEqual(reactor.eval(), aspen.State.COMPLETE_EVALUATED)

  def test_evaluation_without_a_value(self):
    queue = aspen.Queue()
    queue.push(1)
    queue.push(2)
    reactor = aspen.lift(
      lambda value: aspen.FunctionEvaluation(aspen.State.NONE)
        if value == 1 else value, queue)
    self.assertFalse(aspen.has_evaluation(reactor.commit(0)))
    self.assertTrue(aspen.has_evaluation(reactor.commit(1)))
    self.assertEqual(reactor.eval(), 2)

  def test_evaluation_with_a_state(self):
    reactor = aspen.lift(
      lambda value: aspen.FunctionEvaluation(value * 2, aspen.State.COMPLETE),
      aspen.constant(21))
    self.assertEqual(reactor.commit(0), aspen.State.COMPLETE_EVALUATED)
    self.assertEqual(reactor.eval(), 42)

  def test_evaluating_state_without_a_value(self):
    with self.assertRaises(ValueError):
      aspen.FunctionEvaluation(aspen.State.EVALUATED)

  def test_raising_evaluation(self):
    reactor = aspen.lift(failing, aspen.constant(1))
    self.assertTrue(aspen.has_evaluation(reactor.commit(0)))
    with self.assertRaises(ValueError):
      reactor.eval()

  def test_repeated_raising_evaluation(self):
    reactor = aspen.lift(failing, aspen.constant(1))
    reactor.commit(0)
    with self.assertRaises(ValueError):
      reactor.eval()
    with self.assertRaises(ValueError):
      reactor.eval()

  def test_shared_raising_evaluation(self):
    source = aspen.lift(failing, aspen.constant(1))
    left = aspen.lift(lambda value: value, source)
    right = aspen.lift(lambda value: value, source)
    left.commit(0)
    right.commit(0)
    with self.assertRaises(ValueError):
      left.eval()
    with self.assertRaises(ValueError):
      right.eval()


if __name__ == '__main__':
  unittest.main()
